### PR TITLE
Add Xoroshiro128Plus

### PR DESCRIPTION
**Pull request by n8sh**

### CODE REVIEW

**Comment by n8sh** [comment]
The PCG package also has generators with 128 bits of state that produce 64 bits of output, but they require `ucent` and 128-bit multiplication which I don't think any D compilers currently support.

File: source/mir/random/engine/xorshift.d [hunk 1]
```diff
@@ -388,3 +388,91 @@ alias Xorshift1024StarPhi = XorshiftStarEngine!(ulong,1024,31,11,30,0x9e3779b97f
...
+        s1 ^= s0;
+        s[0] = rol!(55,ulong)(s0) ^ s1 ^ (s1 << 14); // a, b
+        s[1] = rol!(36,ulong)(s1); // c
+
+        return result;
+    }
+}
+
+///
+unittest
```
Comment (9il) [inline source/mir/random/engine/xorshift.d]: Version for Mir random tests

**Comment by n8sh** [comment]
Restricted unittest to version(mir_random_test).

File: source/mir/random/engine/xorshift.d [hunk 2]
```diff
@@ -388,3 +388,91 @@ alias Xorshift1024StarPhi = XorshiftStarEngine!(ulong,1024,31,11,30,0x9e3779b97f
...
+        return result;
+    }
+}
+
+///
+@nogc nothrow pure @safe version(mir_random_test) unittest
+{
+    import mir.random.engine : isSaturatedRandomEngine;
+    static assert(isSaturatedRandomEngine!Xoroshiro128Plus);
+    auto gen = Xoroshiro128Plus(1234u);//Seed with constant.
```
Comment (9il) [inline source/mir/random/engine/xorshift.d]: Please test that the first and say 10th generated values maches with the reference implementation (in documented unittest)

File: source/mir/random/engine/xorshift.d [hunk 3]
```diff
@@ -388,3 +388,91 @@ alias Xorshift1024StarPhi = XorshiftStarEngine!(ulong,1024,31,11,30,0x9e3779b97f
...
+
+    @disable this();
+    @disable this(this);
+
+    /// Constructs an $(D Xoroshiro128Plus) generator seeded with $(D_PARAM x0).
+    this()(ulong x0) @nogc nothrow pure @safe
+    {
+        //Seed using splitmix64 as recommended by Vigna.
+        //http://xoroshiro.di.unimi.it/splitmix64.c
+        foreach (ref e; s)
```
Comment (9il) [inline source/mir/random/engine/xorshift.d]: looks strange, s[0] == s[1]

Comment (n8sh) [reply source/mir/random/engine/xorshift.d]: I am not sure what you mean.

File: source/mir/random/engine/xorshift.d (hunk 2 above)
Comment (n8sh) [reply source/mir/random/engine/xorshift.d]: Added.

File: source/mir/random/engine/xorshift.d (hunk 3 above)
Comment (9il) [reply source/mir/random/engine/xorshift.d]: Oh, sorry, I was wrong , did not see `x0 += .. `

File: source/mir/random/engine/xorshift.d [hunk 4]
```diff
@@ -388,3 +388,93 @@ alias Xorshift1024StarPhi = XorshiftStarEngine!(ulong,1024,31,11,30,0x9e3779b97f
...
+        auto s1 = s[1];
+        immutable result = s0 + s1;
+
+        s1 ^= s0;
+        s[0] = rol!(55,ulong)(s0) ^ s1 ^ (s1 << 14); // a, b
+        s[1] = rol!(36,ulong)(s1); // c
+
+        return result;
+    }
+}
```
Comment (9il) [inline source/mir/random/engine/xorshift.d]: Maybe `jump` from http://xoroshiro.di.unimi.it/xoroshiro128plus.c is good to be added too?

Comment (n8sh) [reply source/mir/random/engine/xorshift.d]: Sounds reasonable. Added.

**Pull request merged**